suite/tests/tools.c

```diff
@@ -497,6 +497,86 @@ intercept_signal(int sig, handler_3_t handler, bool sigstack)
     ASSERT_NOERR(rc);
 }
 
+#        ifdef AARCH64
+#            ifdef DR_HOST_NOT_TARGET
+#                define RESERVED __reserved1
+#            else
+#                define RESERVED __reserved
+#            endif
+void
+dump_ucontext(ucontext_t *ucxt, bool is_sve, int vl_bytes)
+{
+    struct _aarch64_ctx *head = (struct _aarch64_ctx *)(ucxt->uc_mcontext.RESERVED);
+    assert(head->magic == FPSIMD_MAGIC);
+    assert(head->size == sizeof(struct fpsimd_context));
+
+    struct fpsimd_context *fpsimd = (struct fpsimd_context *)(ucxt->uc_mcontext.RESERVED);
+    print("\nfpsr 0x%x\n", fpsimd->fpsr);
+    print("fpcr 0x%x\n", fpsimd->fpcr);
+    reinterpret128_2x64_t vreg;
+    int i;
+    for (i = 0; i < MCXT_NUM_SIMD_SVE_SLOTS; i++) {
+        vreg.as_128 = fpsimd->vregs[i];
+        print("q%-2d  0x%016lx %016lx\n", i, vreg.as_2x64.hi, vreg.as_2x64.lo);
+    }
+    print("\n");
+
+    if (is_sve) {
+        size_t offset = sizeof(struct fpsimd_context);
+        struct _aarch64_ctx *next_head =
+            (struct _aarch64_ctx *)(ucxt->uc_mcontext.RESERVED + offset);
+        while (next_head->magic != 0) {
+            switch (next_head->magic) {
+            case ESR_MAGIC: offset += sizeof(struct esr_context); break;
+            case EXTRA_MAGIC: offset += sizeof(struct extra_context); break;
+            case SVE_MAGIC: {
+                const struct sve_context *sve = (struct sve_context *)(next_head);
+                assert(sve->vl == vl_bytes);
+                const unsigned int vq = sve_vq_from_vl(sve->vl);
+                if (sve->head.size != sizeof(struct sve_context))
+                    assert(sve->head.size == ALIGN_FORWARD(SVE_SIG_CONTEXT_SIZE(vq), 16));
+
+                uint64 vdw; /* A vector's doubleword. */
+                int boff;   /* Byte offset for each doubleword in a vector. */
+                for (i = 0; i < MCXT_NUM_SIMD_SVE_SLOTS; i++) {
+                    print("z%-2d  0x", i);
+                    for (boff = ((vq * 2) - 1); boff >= 0; boff--) {
```

Comment:
Same comments as above: I am having a hard time understanding this code.  It looks very similar to the core/ version.  Is it possible to share it?

Reply:
These are separated out is because the intent of each is different. The core's version has extra information about positions and sizes of the context sections in order to debug core defects. The user version in `tools.c` is intended to be used as a utility function to dump just the registers for comparing tests' translated state with native state, so it's simpler.

Also, the core's version writes to DR's log files while `tools.c`'s output is to `stderr`. Although this aspect could be merged, it would complicate the implementation of a common function.

Reply:
This is the sort of logging output we want for debugging in the core using `dump_sigcontext()`:
```
SVE vector length 32 bytes
Quadwords (128 bits) per vector 2

SVE_SIG_ZREG_SIZE 32
SVE_SIG_PREG_SIZE 4
SVE_SIG_FFR_SIZE  4
sve->head.size 16

SVE_SIG_ZREGS_OFFSET 16
SVE_SIG_PREGS_OFFSET 1040
SVE_SIG_FFR_OFFSET   1104
```